### DISCUSSION TITLE
Profile-specific parent profiles?

### DISCUSSION THREAD

**Discussion by edeandrea** [quarkusio/quarkus#24586] [Q&A] [answered]
I was reading through the parent profile stuff in the [Quarkus Configuration Reference](https://quarkus.io/guides/config-reference#parent-profile). I would have expected to be able to assign a parent profile to specific child profiles.

For example, I don't seem to be able to do this:

```
quarkus.http.port=8084
%prod.quarkus.http.port=8080

%foo.quarkus.config.profile.parent=bar
%baz.quarkus.config.profile.parent=prod
```

Then if I built with `-Dquarkus.profile=baz`, I would expect the app's http port to be `8080`, but instead the port is `8084`.

Now, if I create an `application-baz.properties` file with the following content:

```
quarkus.config.profile.parent=prod
```

When I re-build with `-Dquarkus.profile=baz`, I now see the http port at `8080` like I would expect.

Is this a bug? In the main `application.properties` file, shouldn't I be able to assign a specific parent profile only for a certain child profile?

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
Which version are you using? We didn't have support to set the parent profile in a profile itself (if you know what I mean :)). It required to be in the main profile because these configurations are required to set the `Config` so they behave a bit differently.

Anyway, I recently added support for this here: https://github.com/quarkusio/quarkus/pull/24369. So you need Quarkus `2.8.x` or update SmallRye Config manually to `2.9.2`. If it still not working, then it is a bug now.

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
It was `2.7.5.Final`

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
Can you try with the latest `2.8.x` RC?

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
It seems to work the way I would expect it to in `2.8.0.CR1`.

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
Great. Sorry for the inconvenience on the new version :)

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
I noticed on #24369  someone asked about backporting it to 2.7. The only reason I'd suggest it is because Red Hat build of Quarkus 2.7 will be based on 2.7 and this would be a really good thing to have.

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
I guess we can mark it to be backported. Feel free to do it :)

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
i dont know how to do that (nor do I think I have permissions to...). I commented on the PR itself.

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
Is just adding a label to the PR. No worries, I did it.

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
Thanks very much! I'm not special enough to be able to add labels to things :D

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
Hey @radcortez I'm revisiting this & had a question with Quarkus 2.9.2.Final.

If I have something like this in `application.properties`:

```properties
%knative.quarkus.config.profile.parent=prod
%knative.quarkus.kubernetes.deployment-target=knative

%knative-17.quarkus.config.profile.parent=knative
%knative-17.quarkus.docker.dockerfile-jvm-path=src/main/docker/Dockerfile.jvm17

%knative-openshift.quarkus.config.profile.parent=knative
%knative-openshift.quarkus.container-image.builder=openshift

%knative-openshift-17.quarkus.config.profile.parent=knative-openshift
%knative-openshift-17.quarkus.openshift.jvm-dockerfile=src/main/docker/Dockerfile.jvm17
%knative-openshift-17.quarkus.openshift.base-jvm-image=registry.access.redhat.com/ubi8/openjdk-17:1.11
```

This essentially sets up a profile hierarchy

```shell
                     prod
                        |
                  knative
                 /          \
      knative-17      knative-openshift
                                    |
                      knative-openshift-17
```

When I do something like `./mvnw package -Dquarkus.profile=knative-openshift-17` I would have expected the `quarkus.kubernetes.deployment-target=knative` to be "inherited" by the `knative-openshift-17` profile, but it isn't.

Is this kind of hierarchy supported? Or is it really only a single level of parent profiles that is allowed?

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
Hi @radcortez - gentle ping on https://github.com/quarkusio/quarkus/discussions/24586#discussioncomment-2857805

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
Hey @edeandrea,

Sorry, due to JNation, I have been completely away from any coding. Next week I should resume my regular activities. I'm really sorry for the delay.

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
Oh no worries at all!

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
Hey @edeandrea 

Right now, only a single level of the hierarchy is supported. If you feel that we should support multi-level, please open an enhancement request in https://github.com/smallrye/smallrye-config :)

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
Thanks @radcortez I may do just that!

And sorry about hitting you up last week - I completely forgot about how involved you are/were with JNation.

[the remark this answers is not in the document]
**Reply by radcortez** [reply]
No worries. All good :)

[the remark this answers is not in the document]
**Reply by edeandrea** [reply]
Created at smallrye/smallrye-config#767

**Comment by p1ass** [comment]
Does config profile support profile-specific parent profile when we use `quarkus-config-yaml` ?
I reproduce same unexpected behavior on Quarkus `2.9.2`.

**Reply by radcortez** [reply]
Yes.

What behavior?

**Reply by p1ass** [reply]
I wrote config yaml as below:

```yaml
# application.yaml
something-value: "root"
```

```yaml
# application-prod.yaml
something-value: "prod"
```

```yaml
# application-baz.yaml
quarkus:
  config:
    profile:
      parent: prod
```

When I read `something-value` using `@ConfigProperty`, I got `root` but expected is `prod`.

```java
@Path("/hello")
public class GreetingResource {
    @ConfigProperty(name = "something-value") String somethingValue;
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return somethingValue;
    }
}
```

```bash
quarkus dev -Dquarkus.profile=baz

....
2022-06-21 22:21:58,825 INFO  [io.quarkus] (Quarkus Main Thread) Profile baz activated. Live Coding activated.
...
```
```bash
$ curl -s localhost:8080/hello                                                    
root

```

Sample repo is here: https://github.com/p1ass/quarkus-parent-profile-sample

**Reply by radcortez** [reply]
Yes, this does not work. Currently, profiles need to be resolved before loading any profile-aware file, so the `quarkus.config.profile.parent` needs to be placed in the main `properties` file or passed as a system property. 

```yaml
#application.yaml
"%baz":
  quarkus:
    config:
      profile:
        parent: prod
```

**Reply by edeandrea** [reply]
@radcortez I'm doing this currently (https://github.com/quarkusio/quarkus-super-heroes/blob/b878f7601ec5302b05bcb03ee46a0657dcb27fca/rest-villains/src/main/resources/application.properties#L45-L106) and it does seem to work. 

2 differences though between what I'm doing & what @p1ass is doing:
1. I'm not using yaml
2. I'm keeping everything inside `application.properties` rather than having separate files for each profile, but I did pilot using separate files (& setting parent profiles in each file) and it did seem to work.

Unless of course I was dreaming :)

**Reply by radcortez** [reply]
The parent profile (like the main profile) has to be resolved before loading the profile-aware files. This is by design, to keep the same behaviour as `quarkus.profile` and to avoid corner cases where you may have profile-aware files with circular dependencies with the parent profile.

**Reply by p1ass** [reply]
Thank you for replying. 
I understood `quarkus.profile` behavior and design background. 👍 

I have a suggestion, how about adding a note in the documents that we can't use the parent profile in parent-aware files?

https://quarkus.io/guides/config-reference.html#parent-profile

**Reply by radcortez** [reply]
Can I ask you to submit a PR with it? :)

**Reply by p1ass** [reply]
Sure! I'll send a PR later.

**Reply by radcortez** [reply]
Great! Thanks!

**Reply by p1ass** [reply]
Submitted a pull request https://github.com/quarkusio/quarkus/pull/26350
